sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,387 @@
+device behaviors for AOT compilations for GPU new command line interfaces
+are needed to pass along this information.
+
+| Target | Triple        | Offline Tool   | Option for Additional Args |
+|--------|---------------|----------------|----------------------------|
+| CPU    | spir64_x86_64 | opencl-aot     | `--cpu-tool-arg=<arg>`     |
+| GPU    | spir64_gen    | ocloc          | `--gen-tool-arg=<arg>`     |
+| FPGA   | spir64_fpga   | aoc/opencl-aot | `--fpga-tool-arg=<arg>`    |
+
+*Table: Ahead of Time Info*
+
+To complete the support needed for the various targets using the
+`clang-linker-wrapper` as the main interface, a few additional options will
+be needed to communicate from the driver to the tool.  Further details of usage
+are given further below.
+
+| Option Name                  | Purpose                                      |
+|------------------------------|----------------------------------------------|
+| `--fpga-link-type=<arg>`     | Tells the link step to perform 'early' or 'image' processing to create archives for FPGA |
+| `--parallel-link-sycl=<arg>` | Provide the number of parallel jobs that will be used when processing with `llvm-foreach` |
+| `--no-sycl-device-lib=<arg>` | Provide the list of device libraries to restrict from linking during device link |
+
+*Table: Additional Options for clang-linker-wrapper*
+
+The `clang-linker-wrapper` provides an existing option named `-wrapper-jobs`
+that may be useful for our usage instead of creating a new option specific
+to `llvm-foreach` processing.
+
+#### spir64_gen support
+
+Compilation behaviors involving AOT for GPU involve an additional call to
+the OpenCL Offline compiler (OCLOC).  This call occurs after the post-link
+step performed by sycl-post-link.  Additional options passed by the user
+via the `-Xsycl-target-backend=spir64_gen <opts>` command as well as the
+implied options set via target options such as `-fsycl-targets=intel_gpu_skl`
+will be processed by a new options to the wrapper, `--gen-tool-arg=<arg>`
+
+To support multiple target specifications, for instance:
+`-fsycl-targets=intel_gpu_skl,intel_gpu_pvc`, multiple `--gen-tool-arg`
+options can be passed on the command line.  Each instance will be considered
+a separate OCLOC call passing along the `<args>` as options to the OCLOC call.
+The compiler driver will be responsible for putting together the full option
+list to be passed along.
+
+> -fsycl -fsycl-targets=spir64_gen,intel_gpu_skl
+-Xsycl-target-backend=spir64_gen "-device pvc -options -extraopt_pvc"
+-Xsycl-target-backend=intel_gpu_skl "-options -extraopt_skl"
+
+*Example: spir64_gen enabling options*
+
+> --gen-tool-arg="-device pvc -options extraopt_pvc"
+--gen-tool-arg="-device skl -options -extraopt_skl"
+
+*Example: clang-linker-wrapper options*
+
+Each OCLOC call will be represented as a separate device binary that is
+individually wrapped and linked into the final executable.
```

Comment:
How does `clang-linker-wrapper` know how many times to invoke `ocloc`?  How does it know which GPU target is associated with each `ocloc` invocation?

Is the idea that `clang-linker-wrapper` will count the number of `--gen-tool-arg` options and parse the value to find the "-device pvc" part?

Would it make sense to change the syntax to make it easier to find the device name, something like:

```
--gen-tool-arg=pvc "-options extraopt_pvc"
```

Something else to keep in mind ... my thought is that we will change `sycl-post-link` to emit a separate device image for each GPU target.  This is another reason why `clang-linker-wrapper` will need to know the set of GPU targets.

Reply:
Good point - we will need to provide a way to differentiate more than just the arch (spir64_gen vs spir64_x86_64).  I like your idea of taking the known syntax of `-opt=<target> "opts"` which we currently use for `-Xsycl-target-backend`